cmd/testscript/main.go

```diff
@@ -0,0 +1,249 @@
+		mainUsage(os.Stderr)
+	}
+	flag.Parse()
+
+	td, err := ioutil.TempDir("", "testscript_")
+	if err != nil {
+		return fmt.Errorf("unable to create temp dir: %v", err)
+	}
+	defer os.RemoveAll(td)
+	runDir = td
+
+	files := flag.Args()
+	if len(files) == 0 {
+		files = []string{os.Stdin.Name()}
+	}
+
+	for i, fn := range files {
+		// TODO make running files concurrent by default? If we do, note we'll need to do
+		// something smarter with the runner stdout and stderr below
+		rd := filepath.Join(td, "dir"+strconv.Itoa(i))
+		if err := os.Mkdir(rd, 0777); err != nil {
+			return fmt.Errorf("failed to create a run directory within %v for %v: %v", td, fn, err)
+		}
+		if err := run(rd, fn); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+var failedRun = errors.New("failed run")
+
+type runner struct {
+	verbose bool
+}
+
+func (r runner) Skip(is ...interface{}) {
+	panic("Skip is not supported")
+}
+
+func (r runner) Fatal(is ...interface{}) {
+	r.Log(is...)
+	r.FailNow()
+}
+
+func (r runner) Parallel() {
+	// No-op for now; we are currently only running a single script in a
+	// testscript instance.
+}
+
+func (r runner) Log(is ...interface{}) {
+	fmt.Print(is...)
+}
+
+func (r runner) FailNow() {
+	panic(failedRun)
+}
+
+func (r runner) Run(n string, f func(t testscript.T)) {
+	// For now we we don't top/tail the run of a subtest. We are currently only
+	// running a single script in a testscript instance, which means that we
+	// will only have a single subtest.
+	f(r)
+}
+
+func (r runner) Verbose() bool {
+	return r.verbose
+}
+
+func run(rd, fn string) error {
+	var ar *txtar.Archive
+	var err error
+
+	testdata := filepath.Join(rd, "testdata")
+	mods := filepath.Join(testdata, "goproxytest")
+
+	if err := os.MkdirAll(mods, 0777); err != nil {
+		return fmt.Errorf("failed to create testdata dir: %v", err)
+	}
+
+	if fn == os.Stdin.Name() {
+		byts, err := ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return fmt.Errorf("failed to read from stdin: %v", err)
+		}
+		ar = txtar.Parse(byts)
+	} else {
+		ar, err = txtar.ParseFile(fn)
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to txtar parse %v: %v", fn, err)
+	}
+
+	var toWrite []txtar.File
+
+	for _, f := range ar.Files {
+		fp := filepath.Clean(filepath.FromSlash(f.Name))
+		if isAbs(fp) || strings.HasPrefix(fp, ".."+string(filepath.Separator)) {
+			return fmt.Errorf("file %v would result in file outside parent directory", f.Name)
+		}
+
+		// only write out goproxytest rooted files
+		if strings.Split(fp, string(os.PathSeparator))[0] != "goproxytest" {
+			toWrite = append(toWrite, f)
+			continue
+		}
+
+		fp = filepath.Join(testdata, fp)
+
+		if err := os.MkdirAll(filepath.Dir(fp), 0777); err != nil {
+			return err
+		}
+		// Avoid overwriting existing files by using O_EXCL.
+		out, err := os.OpenFile(fp, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0666)
+		if err != nil {
+			return err
+		}
+
+		// unquote, because this will be a quoted goproxytest file
```

Comment:
I don't think this is necessary any more.

Reply:
Yes, yet another thing I missed. Thanks for the catch.